Store TxHistory outside of the wallet state

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#90

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have moved tracking of `Tx` outside of the wallet state (which does now only keep track of the known TxId and TxMeta)
- [x] I have extended the DB Layer with two new: `putTxHistory` and `readTxHistory`
- [x] I have extended the MVar spec accordingly with a few properties to reflect upon the semantic of the DB Layer
- [x] I have reviewed and extended a few docstring comments.

# Comments

<!-- Additional comments or screenshots to attach if any -->

This is because transactions may end up being rather big and, are of little use for the wallet
state. Instead, keeping track of the tx ids only is sufficient. Storing transactions on a
separate level in the DB also allows for querying them independently with, later, more advanced
features like sorting and filtering

NOTE: More tests on the wallet model are introduced by @rvl in #148 so I really focused on testing the DB layer here.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

akegalj commented on .github/ISSUE_TEMPLATE/adr.md:
```diff
@@ -6,35 +6,64 @@ about: Document an Architectural Decision
+- [ ] I intend to extend the existing handlers and use the wallet layer to implement 
+  the necessary steps.
+
+- [ ] I plan on testing the endpoint by adding a few integration scenarios
```
Aha - I was thinking on more detailed plan scenario like:
 * I intend to modify function at L222 to take extra argument `Foo`
```haskell
L222: boo :: a -> b -> Foo -> c
```

-----

akegalj commented on .github/ISSUE_TEMPLATE/adr.md:
```diff
@@ -43,19 +72,37 @@ e.g.: The API _must_ support creation of wallets through a dedicated endpoint.>
+
+| Criteria | Coverage |
+| ---      | ---      |
+| 1.       | The creation has been covered has part of testing (cf Scenario/Transactions, tests prefixed with #14) | 
```
I gess you though 
> The creation has been covered as part of testing 

s/has part/as part/

KtorZ committed:
handle tx history separately from the wallet state

This is because transactions may end up being rather big and, are of little use for the wallet
state. Instead, keeping track of the tx ids only is sufficient. Storing transactions on a
separate level in the DB also allows for querying them independently with, later, more advanced
features like sorting and filtering

KtorZ committed:
extend MVarSpec to cover new putTxHistory & getTxHistory interfaces

paweljakubas commented on src/Cardano/Wallet/DB/MVar.hs:
```diff
@@ -15,25 +17,52 @@ module Cardano.Wallet.DB.MVar
+
         , readCheckpoint = \key ->
-            Map.lookup key <$> readMVar wallets
+            fmap fst . Map.lookup key <$> readMVar db
```
ok, so basically we are storing the pair (wallet state, txs) and keep it separated. And we have separate db calls for the first one and transactions part. And txs history is not polluting and affecting wallet state -> straightforward property to test

paweljakubas commented on src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -103,7 +111,7 @@ data Wallet s where
         => UTxO -- Unspent tx outputs belonging to this wallet
         -> Set Tx -- Pending transactions
-        -> Set (Tx, TxMeta) -- Transaction history
+        -> Set (Hash "Tx", TxMeta) -- Transaction history
```
right, we probably do not need to store whole transaction content, its identifier is enough

KtorZ replied:
In theory, I think we don't even need to store the `TxMeta`. Not sure why I kept them here :thinking: ...

paweljakubas commented on test/unit/Cardano/Wallet/DB/MVarSpec.hs:
```diff
@@ -47,6 +83,12 @@ spec = do
+            \db -> (checkCoverage $ dbMergeTxHistoryProp db)
+        it "can't read Tx history if there's no checkpoint" $
+            \db -> (property $ dbPutTxHistoryBeforeCheckpointProp db)
+        it "putTxHistory leaves the wallet state untouched" $
```
oh here is the property I thought immediately when got idea behind decoupling of wallet state and txs :100:   but there is symmetry to be tested, ie., also on the contrary - changing the state leaves TxHistory unaffected, right?

KtorZ replied:
> changing the state leaves TxHistory unaffected, right?

Indeed. I've got a bit lazy here :sweat_smile: I'll add.

paweljakubas commented on test/unit/Cardano/Wallet/DB/MVarSpec.hs:
```diff
@@ -71,44 +112,84 @@ dbReplaceValsProp db (key, val1, val2)  = monadicIO $ liftIO $ do
+        sumOfSizes k = sum $ Map.size . snd <$> restrictTo k keyValPairs
+    prop = monadicIO $ liftIO $ do
+        forM_ keyValPairs $ \(key, val) -> do
+            putCheckpoint db key (initWallet $ DummyState 0)
```
ok, so we are here putting tx histories using between (10,50) pairs, let's say X and having X wallet state checkpoints, each having `DummyState 0`. Then, for each key form X we read tx history. And expect that at each key the corresponding value is the same from res and when looking at keyValPairs (here Map.unions gather values if the key occurs more than once)

paweljakubas commented on test/unit/Cardano/Wallet/DB/MVarSpec.hs:
```diff
@@ -71,44 +112,84 @@ dbReplaceValsProp db (key, val1, val2)  = monadicIO $ liftIO $ do
+    restrictTo k = filter ((== k) . fst)
+    -- Make sure that we have some conflicting insertion to actually test the
+    -- semantic of the DB Layer.
+    cond = map sumOfSizes ids /= map sizeOfUnion ids
```
so in 90% of cases we have the situation that the same key occurred more than once in keyValPairs and hence the values coalesced. If yes, can't we just put :  
```
(L.length . L.nub) ids < L.length ids
``` 
?

KtorZ replied:
Hmmm.... :thinking: It's true that we only get conflict if we have the same wallet ids and some similar tx ids. So it should be enough indeed to just look at the ids and not at the balance to know whether we're going to have conflict. The balance being different is just a consequence of the id being in conflict. Good point. I'll simplify that :+1:

paweljakubas reviewed (approved):
LGTM
